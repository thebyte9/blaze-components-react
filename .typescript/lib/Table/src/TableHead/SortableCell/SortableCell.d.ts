/// <reference types="react" />
declare const SortableCell: ({ onSort, orderBy, column, columns, appliedSort }: {
    onSort: any;
    orderBy: any;
    column: any;
    columns: any;
    appliedSort: any;
}) => JSX.Element;
export default SortableCell;
