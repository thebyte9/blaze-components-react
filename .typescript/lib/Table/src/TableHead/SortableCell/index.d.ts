import SortableCell from "./SortableCell";
export default SortableCell;
