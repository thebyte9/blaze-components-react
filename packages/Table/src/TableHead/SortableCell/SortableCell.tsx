import React, { useEffect, useState } from "react";

const SortableCell = ({
  onSort,
  orderBy,
  column,
  columns,
  appliedSort
}: {
  onSort: any;
  orderBy: any;
  column: any;
  columns: any;
  appliedSort: any;
}) => {
  const formatColumns = columns.reduce((acc: any, item: any): any => {
    return { ...acc, [item]: null };
  }, {});
  const [tableColumns, setTableColumns] = useState(formatColumns);

  type TSortDirection = "asc" | "desc" | null;
  const asc: TSortDirection = "asc";
  const desc: TSortDirection = "desc";
  const hide: TSortDirection = null;

  const getSortDirection = (col: string): TSortDirection => {
    if (tableColumns[col] === hide) {
      return asc;
    }
    return tableColumns[col] === asc ? desc : asc;
  };

  const sort = (col: any) => {
    if (!orderBy.includes(col)) {
      return;
    }

    const resetTableColumns = Object.keys(tableColumns).reduce(
      (acc: any, key: any) => {
        acc[key] = hide;
        return acc;
      },
      {}
    );

    const sortDirection = getSortDirection(column);

    setTableColumns({
      ...resetTableColumns,
      [column]: sortDirection
    });

    onSort({
      [column]: sortDirection
    });
  };

  useEffect(() => {
    if (appliedSort) {
      const [[col, direction]] = appliedSort && Object.entries(appliedSort);
      if (tableColumns[col] !== direction) {
        const merged = {
          ...tableColumns,
          [col]: "asc"
        };

        setTableColumns(merged);
      }
    }
  }, [appliedSort]);

  return (
    <div className="sortable">
      <span
        data-testid={`sortby-${column}`}
        onClick={() => sort(column)}
        role="button"
      >
        {column}
      </span>

      {tableColumns[column] !== hide && (
        <i className="material-icons">
          {tableColumns[column] === asc
            ? "keyboard_arrow_up"
            : "keyboard_arrow_down"}
        </i>
      )}
    </div>
  );
};

export default SortableCell;
