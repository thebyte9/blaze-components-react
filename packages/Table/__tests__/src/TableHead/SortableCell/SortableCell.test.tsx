import { cleanup, fireEvent, render } from "@testing-library/react";
import expect from "expect";
import React from "react";
import SortableCell from "../../../../src/TableHead/SortableCell/SortableCell";
import { data } from "../../../mocks";

describe("Sortable cell", () => {
  afterEach(cleanup);

  it("should be defined", () => {
    expect(SortableCell).toBeDefined();
  });

  it("should render without throwing error", () => {
    const { container } = render(
      <SortableCell
        appliedSort={null}
        onSort={jest.fn()}
        orderBy={data.orderBy}
        column={data.columns[0]}
        columns={data.columns}
        labels={data.labels}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should render without throwing error", () => {
    const onSortMock = jest.fn();

    const { getByTestId } = render(
      <SortableCell
        appliedSort={{ name: "asc" }}
        onSort={onSortMock}
        orderBy={data.orderBy}
        column={data.columns[1]}
        columns={data.columns}
        labels={data.labels}
      />
    );

    const sort = getByTestId(`sortby-${data.columns[1]}`);

    fireEvent.click(sort);

    expect(onSortMock).toHaveBeenCalledWith({ [data.columns[1]]: "desc" });
  });
});
