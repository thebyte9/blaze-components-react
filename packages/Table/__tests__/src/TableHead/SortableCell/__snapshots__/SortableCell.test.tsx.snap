// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sortable cell should render without throwing error 1`] = `
<div>
  <div
    class="sortable"
  >
    <span
      data-testid="sortby-id"
      role="button"
    >
      id
    </span>
  </div>
</div>
`;
